Adding streaming support through Jetpack API

joelclimbsthings opened the pull request:
- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR switches over to using the Jetpack `jetpack-ai-query` streaming endpoint for product descriptions, and implementing that within the editor. It utilizes some utilities copied from the Jetpack AI block, and I created the `useCompletion` hook to simplify it's usage.

This also removes the need for a fair amount of code, perhaps including our own REST endpoints but I've left those just in case @nima-karimi 's use-case requires it. 

Closes #38413 .

### Screenshots
![image](https://github.com/woocommerce/woocommerce/assets/444632/2ecf7f16-c521-42c0-b17d-14185cf9b5e9)


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. I've been using this with a WoA dev site ([guide](https://explorersp2.wordpress.com/2021/11/08/wordpress-on-atomic-development-workflow/)). Following that guide, but with [this modified script](https://gist.github.com/joelclimbsthings/d0144f525e165d4943650feeec09c9a4).
2. Once you've copied the plugin over, ensure the `Woo AI` plugin is activated.
3. Go to Products -> Add new
4. You should see the "Write it for me" button as pictured above.
5. Observe that it's disabled until you enter a Product title of ~20 characters.
6. After entering a title, it should be enabled.
7. After clicking the button, you should see the familiar streaming behavior within the description editor creating your description.
8. The button with change to "Stop writing"
9. Ensure the the completion finishes without an issue.
10. Then try again, and hit "Stop writing" to ensure that the streaming stops as expected.

<!-- End testing instructions -->

joelclimbsthings committed:
Adding streaming support through Jetpack API

joelclimbsthings committed:
Preventing displaying of incomplete HTML tags

joelclimbsthings commented on plugins/woo-ai/src/product-description/product-description-button-container.tsx:
```diff
@@ -1,61 +1,68 @@
 
-	const writeItForMeDisabled =
-		fetching || ! productTitle || productTitle.length < MIN_TITLE_LENGTH;
+				tinyEditor.setContent( getApiError() );
```
This error handling isn't great at the moment, but I'm uncertain of what the endpoint actually supports (and it sounds like not much so far).

tommyshellberg replied:
Yeah, we may end up changing this anyways to use Gutenberg notices/snackbars once we start to improve the UX.

joelclimbsthings committed:
Load Jetpack class conditionally

joelclimbsthings committed:
Making description component render conditional

joelclimbsthings committed:
Fixing typo

joelclimbsthings committed:
Preventing error when openai key is not defined

joelclimbsthings committed:
Fixing typo

tommyshellberg commented:
Bravo for getting streaming working for the product descriptions! It feels so much faster than before, definitely a big improvement and I'm glad we looked into it. 👍 
I will scour the code some more in the morning but one thing I wonder about is if we could set a fixed with for the "write it for me" and "stop writing" button. It's not a deal-breaker but if we could avoid a layout shift when the button changes, that would be ideal.

joelclimbsthings committed:
Adding min width to write it for me button

joelclimbsthings commented:
Great thought @tommyshellberg , and thanks for checking it out! I've addressed the button width in https://github.com/woocommerce/woocommerce/pull/38407/commits/8ca15e171993a07931be26c31c9dc3ce302e11d3 .

I'm a little uncertain of whether to keep the icon on the button for consistency, but I've also added that for now.

tommyshellberg commented:
What do we think about including your atomic-watch script in the repo, say, in the `bin/` directory (which will be excluded from the build ZIP plugin)? I know that it's not designed to be executed directly using tooling/automation but it seems like it should exist somewhere in the repo so that other folks have access to it should someone jump into the project. 🤔 
An alternative would be to create a follow-up issue to generate a DEVELOPMENT.md file with a short set of instructions on how to set up syncing using the script.

tommyshellberg commented on plugins/woo-ai/src/product-description/product-description-button-container.tsx:
```diff
@@ -93,52 +92,24 @@ export function WriteItForMeButtonContainer() {
-			'woocommerce'
+	if ( completionActive ) {
+		return (
+			<button
```
Nitpick: should this be its own component within /product-description/ ?

tommyshellberg commented on plugins/woo-ai/src/product-description/product-description-button-container.tsx:
```diff
@@ -93,52 +92,24 @@ export function WriteItForMeButtonContainer() {
+	}
 
 	return (
 		<button
```
Same as above, should we put these two buttons into their own components?
IMO it can be faster to skim this file if it looks something like:

```
	{
		completionActive ? (
			<StopCompletionButton {props} />
		) : (
			<GenerateCompletionButton  {props}/>
		)
	}
}
```
I know it's not fun passing down multiple props (although you've created a hook which makes it easier) so I'll defer to your React expertise on whether separating them out into components for readability is worth it.

joelclimbsthings replied:
I was a little on the fence about this, but gave it a go in https://github.com/woocommerce/woocommerce/pull/38407/commits/20b37d365f2618c96ea84d125a09c51204af7f5a. I think it's a bit cleaner, and I just left a couple props (onClick/disabled) in the root file to prevent too much prop drilling. 👍🏻

joelclimbsthings committed:
Refactoring completion buttons

joelclimbsthings commented:
Thanks for the review @tommyshellberg !

> What do we think about including your atomic-watch script in the repo, say, in the bin/ directory 

I like the idea of having this accessible, although that script is intended for use with a WoA and I don't believe those are public. Also, the script really isn't specific to our plugin exactly, as it could be run in any plugin directory to copy it over. 

All that said, maybe we could handle that type of task in another issue specific to docs?

tommyshellberg commented:
> I like the idea of having this accessible, although that script is intended for use with a WoA site and I don't believe those are public? Also, the script really isn't specific to our plugin exactly, as it could be run in any plugin directory to copy it over.
> All that said, maybe we could handle that type of task in another issue specific to docs?

That works; we can create an issue for adding basic docs. 👍

tommyshellberg reviewed (approved):
Great work here, and thanks for responding to the feedback. 👍 
I have just the smallest of questions but this is ready to merge IMO. 🚢

tommyshellberg commented:
I approved this but see now that Nima's title suggestions PR has merged there are conflicts. 
Happy to review this again if you'd like after the conflicts are settled.

joelclimbsthings committed:
Simplifying handler
